src/mlpack/core/tree/address.hpp

```diff
@@ -0,0 +1,249 @@
+ * variables should be equal-sized and the type of the address should correspond
+ * to the type of the vector.
+ *
+ * @param address An address to translate.
+ * @param point The point that corresponds to the address.
+ */
+template<typename AddressType, typename VecType>
+void AddressToPoint(VecType& point, const AddressType& address)
+{
+  typedef typename VecType::elem_type VecElemType;
+  typedef typename AddressType::elem_type AddressElemType;
+
+  constexpr size_t order = sizeof(AddressElemType) * CHAR_BIT;
+  // Calculate the number of bits for the exponent.
+  const int numExpBits = std::ceil(std::log2(
+      std::numeric_limits<VecElemType>::max_exponent -
+      std::numeric_limits<VecElemType>::min_exponent + 1.0));
+
+  assert(point.n_elem == address.n_elem);
+  assert(address.n_elem > 0);
+
+  arma::Col<AddressElemType> rearrangedAddress(address.n_elem,
+      arma::fill::zeros);
+  // Calculate the number of bits for the mantissa.
+  const int numMantBits = order - numExpBits - 1;
+
+  for (size_t i = 0; i < order; i++)
+    for (size_t j = 0; j < address.n_elem; j++)
+    {
+      size_t bit = (i * address.n_elem + j) % order;
+      size_t row = (i * address.n_elem + j) / order;
+
+      rearrangedAddress(j) |= (((address(row) >> (order - 1 - bit)) & 1) <<
+          (order - 1 - i));
+    }
+
+  for (size_t i = 0; i < rearrangedAddress.n_elem; i++)
+  {
+    bool sgn = rearrangedAddress(i) & ((AddressElemType) 1 << (order - 1));
+
+    if (!sgn)
+    {
+      rearrangedAddress(i) = ((AddressElemType) 1 << (order - 1)) - 1 -
+          rearrangedAddress(i);
+    }
+
+    // Extract the mantissa.
+    AddressElemType tmp = (AddressElemType) 1 << numMantBits;
+    AddressElemType mantissa = rearrangedAddress(i) & (tmp - 1);
+    if (mantissa == 0)
+      mantissa = 1;
+
+    VecElemType normalizedVal = (VecElemType) mantissa / tmp;
+
+    if (!sgn)
+      normalizedVal = -normalizedVal;
+
+    // Extract the exponent
+    tmp = (AddressElemType) 1 << numExpBits;
+    AddressElemType e = (rearrangedAddress(i) >> numMantBits) & (tmp - 1);
+
+    e += std::numeric_limits<VecElemType>::min_exponent;
+
+    point(i) = std::ldexp(normalizedVal, e);
+    if (std::isinf(point(i)))
+    {
+      if (point(i) > 0)
+        point(i) = std::numeric_limits<VecElemType>::max();
+      else
+        point(i) = std::numeric_limits<VecElemType>::lowest();
+    }
+  }
+}
+
+/**
+ * Compare two addresses. The function returns 1 if the first address is greater
+ * than the second one, -1 if the first address is less than the second one,
+ * otherwise the function returns 0.
+ */
+template<typename AddressType1, typename AddressType2>
+int CompareAddresses(const AddressType1& addr1, const AddressType2& addr2)
+{
+  static_assert(sizeof(typename AddressType1::elem_type) ==
+      sizeof(typename AddressType2::elem_type), "We aren't able to compare "
+      "adresses of distinct sizes");
```

Comment:
The type of an address is either `uint32_t` or `uint64_t`
```
typedef typename std::conditional<sizeof(ElemType) * CHAR_BIT <= 32,
                                    uint32_t,
                                    uint64_t>::type AddressElemType;
```
So, there is no difference. But maybe you are right and it is better to check types instead of sizes.